account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py

```diff
@@ -0,0 +1,232 @@
+                if is_scheduled:
+                    # Handle all stop conditions for scheduled runs.
+                    if identifier == self.ponto_last_identifier or (
+                        not self.ponto_last_identifier
+                        and transaction_datetime < date_stop
+                    ):
+                        return lines
+                else:
+                    # Handle stop conditions for non scheduled runs.
+                    if transaction_datetime < date_stop:
+                        return lines
+                    if (
+                        transaction_datetime < date_since
+                        or transaction_datetime > date_until
+                    ):
+                        continue
+                line["transaction_datetime"] = transaction_datetime
+                lines.append(line)
+            latest_identifier = transactions[-1].get("id")
+            transactions = interface_model._get_transactions(
+                access_data, latest_identifier
+            )
+        # We get here if we found no transactions before date_since,
+        # or not equal to stored last identifier.
+        return lines
+
+    def _ponto_store_lines(self, lines):
+        """Store transactions retrieved from Ponto in statements.
+
+        The data retrieved has the most recent first. However we need to create
+        the bank statements in ascending date order. as the balance_end of
+        one statement will be the balanxe_start of the next statement.
+        """
+
+        def reset_transactions(date_since):
+            """Reset values for new statement."""
+            statement_date_since = self._get_statement_date_since(date_since)
+            statement_date_until = (
+                statement_date_since + self._get_statement_date_step()
+            )
+            statement_lines = []
+            return statement_date_since, statement_date_until, statement_lines
+
+        lines = sorted(lines, key=itemgetter("transaction_datetime"))
+        (
+            statement_date_since,
+            statement_date_until,
+            statement_lines,
+        ) = reset_transactions(lines[0]["transaction_datetime"])
+        for line in lines:
+            line.pop("transaction_datetime")
+            vals_line = self._ponto_get_transaction_vals(line)
+            if vals_line["date"] >= statement_date_until:
+                self._create_or_update_statement(
+                    (statement_lines, {}), statement_date_since, statement_date_until
+                )
+                (
+                    statement_date_since,
+                    statement_date_until,
+                    statement_lines,
+                ) = reset_transactions(statement_date_until)
+            statement_lines.append(vals_line)
+        # Handle lines in last statement.
+        self._create_or_update_statement(
+            (statement_lines, {}), statement_date_since, statement_date_until
+        )
```

Comment:
After testing this module with real data, we noticed that a few lines were retrieved from ponto but were not imported in odoo. The reason seems to be that the date of the lines not always falls back into the date range (date_since - date_until) and so they are filtered out. I had difficulties to fix this method so I ended up with rewriting this method and simplifying the code this way:



```python
    def _ponto_store_lines(self, lines):
        """Store transactions retrieved from Ponto in statements.
        """
        lines = sorted(lines, key=itemgetter("transaction_datetime"))

        # Group statement lines by date per period (date range)
        grouped_periods = {}
        for line in lines:
            date_since = line["transaction_datetime"]
            statement_date_since = self._get_statement_date_since(date_since)
            statement_date_until = statement_date_since + self._get_statement_date_step()
            if (statement_date_since, statement_date_until) not in grouped_periods:
                grouped_periods[(statement_date_since, statement_date_until)] = []

            line.pop("transaction_datetime")
            vals_line = self._ponto_get_transaction_vals(line)
            grouped_periods[(statement_date_since, statement_date_until)].append(vals_line)

        # For each period, create or update statement lines
        for period, statement_lines in grouped_periods.items():
            (statement_date_since, statement_date_until) = period

            self._create_or_update_statement(
                (statement_lines, {}), statement_date_since, statement_date_until
            )
```

Reply:
@astirpe Hi, could you make a PR along these lines to my branch? Just to be sure I updated things like intended. Anyway, thank you for your contribution!